File: LC014LongestCommonPrefix.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/longest-common-prefix/
 */
public class LC014LongestCommonPrefix {
    public class Solution {
        public String longestCommonPrefix(String[] strs) {
            if (strs == null || strs.length == 0) {
                return "";
            }
            int i = 0;
            boolean done = false;
            while (!done) {
                for (String str : strs) {
                    if (str == null) {
                        return "";
                    }
                    if (i >= str.length() || str.charAt(i) != strs[0].charAt(i)) {
                        done = true;
                        break;
                    }
                }
                i++;
            }
            return strs[0].substring(0, i - 1);
        }
    }

    @Test
    public void test_Ala_AlaMaKota() throws Exception {
        assertEquals("Ala", new Solution().longestCommonPrefix(new String[]{"Ala", "Ala Ma Kota"}));
    }

    @Test
    public void test_aa_a() throws Exception {
        assertEquals("a", new Solution().longestCommonPrefix(new String[]{"aa", "a"}));
    }

    @Test
    public void test_None() throws Exception {
        assertEquals("", new Solution().longestCommonPrefix(new String[]{}));
    }

    @Test
    public void test_Empty_b() throws Exception {
        assertEquals("", new Solution().longestCommonPrefix(new String[]{"", "b"}));
    }
}

File: algorithm-design/leetcode/lc014_longest_common_prefix/LC014LongestCommonPrefix.cs
// https://leetcode.com/problems/longest-common-prefix/

namespace AlgorithmDesign.leetcode.lc014_longest_common_prefix
{
    public class Solution
    {
        public string LongestCommonPrefix(string[] strs)
        {
            if (strs == null || strs.Length == 0)
            {
                return "";
            }
            int i = 0;
            bool done = false;
            while (!done)
            {
                foreach (string str in strs)
                {
                    if (str == null)
                    {
                        return "";
                    }
                    if (i >= str.Length || str[i] != strs[0][i])
                    {
                        done = true;
                        break;
                    }
                }
                i++;
            }
            return strs[0].Substring(0, i - 1);
        }
    }
}
